lib/js/frame.js

```diff
@@ -0,0 +1,72 @@
+var Frame = function(game) {
+  this.game = game;
+  this.total = 0;
+  this.rolls = 0;
+  this.bonusRolls = 0;
+};
+
+Frame.prototype.accepts = function(value) {
+  var pinsRemaining = 10 - (this.total % 10);
+  if (value > pinsRemaining) return false;
+  if (value < 0) return false;
+  if (!Number.isInteger(value)) return false;
+  return true;
+};
+
+Frame.prototype.add = function(value) {
+  this.total += value;
+};
+
+Frame.prototype.roll = function(value) {
+  this._expendBonuses(value);
+  this.add(value);
+  this.rolls ++;
+  if (this.total === 10) this._setBonus();
+};
+
+Frame.prototype.isComplete = function() {
+  if (this._isFinal()) return this._tenthFrameComplete();
+  else return this._normalComplete();
+};
+
+Frame.prototype.hasBonus = function() {
+  return this.bonusRolls > 0;
+};
+
+Frame.prototype.expendBonus = function(value) {
+  if (this.hasBonus()) {
+    this._reduceBonus();
+    if (!this._isFinal()) this.add(value);
+  }
+};
+
+Frame.prototype._expendBonuses = function(value) {
+  this.game.frames.forEach(function(frame) {
+    frame.expendBonus(value);
+  });
+};
+
+Frame.prototype._setBonus = function() {
+  if (this.rolls == 1) this.bonusRolls += 2; // strike
+  else this.bonusRolls += 1; // spare
+};
+
+Frame.prototype._reduceBonus = function() {
+  this.bonusRolls -= 1;
+};
+
+Frame.prototype._isFinal = function() {
+  return this === this.game.frames[9];
+};
+
+Frame.prototype._normalComplete = function() {
+  return this.rolls >= 2 || this.total >= 10;
+};
+
+Frame.prototype._tenthFrameComplete = function() {
+  if (this.rolls >= 3) return true;
+  if (this.rolls == 2 && this.bonusRolls == 0) return true;
+  return false;
+};
+
+module.exports = Frame;
```

Comment:
+ The use of private methods for readability here – it makes the Frame much easier to understand. 
+ The way your conditionals read. You put the simple case first, making it easy to understand, and make good use of JavaScript's sugary syntax to omit unnecessary brackets. It's easy to read.

- I wonder if some of your conditionals could be extracted to statements. Something like

```javascript
if (this.rolls >=3) return true;
```

Could have the conditional (the bit in the brackets) extracted like this:

```javascript
var finishedRolling = this.rolls >=3 || (this.rolls == 2 && this.bonusRolls == 0)

if(finishedRolling) return true;
```

And finally, don't forget that whenever you find yourself writing `if(condition) true else false`, you can just `return condition`:

```javascript
var finishedRolling = this.rolls >=3 || (this.rolls == 2 && this.bonusRolls == 0)

return finishedRolling;
```

So, for an even simpler method overall, line 66 could be:

```javascript
Frame.prototype._tenthFrameComplete = function() {
 return this.rolls >=3 || (this.rolls == 2 && this.bonusRolls == 0)
};
```